[Feature] Delete feature (#202)
tidb-incubator/TiBigData: flink/README_unified_batch_streaming.md

## 6 Delete in Streaming Mode

Create source table and target table with same schema in TiDB:

```sql
CREATE TABLE `test`.`src_table`(
    id BIGINT(20) PRIMARY KEY ,
    name VARCHAR(16) ,
    PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
);

CREATE TABLE `test`.`target_table`(
    id BIGINT(20) PRIMARY KEY ,
    name VARCHAR(16) ,
    PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
);
```

Start Flink SQL client in streaming mode, create TiDB Catalog and insert:

```sql
SET 'sql-client.execution.result-mode' = 'table';

CREATE CATALOG `tidb`
WITH (
  'type' = 'tidb',
  'tidb.database.url' = 'jdbc:mysql://localhost:4000/',
  'tidb.username' = 'root',
  'tidb.password' = '',
  'tidb.streaming.source' = 'kafka',
  'tidb.streaming.codec' = 'json',
  'tidb.streaming.kafka.bootstrap.servers' = 'localhost:9092',
  'tidb.streaming.kafka.topic' = 'test_cdc',
  'tidb.streaming.kafka.group.id' = 'test_cdc_group',
  'tidb.streaming.ignore-parse-errors' = 'true',
  'tidb.sink.impl' = 'TIKV',
  'tidb.write_mode' = 'upsert',
  'tikv.sink.transaction' = 'MINIBATCH',
  'tikv.sink.delete_enable' = 'true'
);

> INSERT INTO `tidb`.`test`.`target_table` SELECT id,name FROM `tidb`.`test`.`src_target`;

Reviewer: Is this line necessary? Looks like it's executing batch inserting?

Author: it is necessary. 
first, as for like batch：batch or streaming depends on the config `tidb.streaming.source` in catalog. I don't known if is there any better way like `select from kafka topic`. I only find this way in the doc.
second, as for like insert：flink does not support delete statement. `flink-connector-jdbc` will also delete with insert  statement

```

> Insert into src table, and you will find data also be inserted into target table

Reviewer: ```suggestion
Insert into source table, and you will find data also be inserted into target table
```

```sql

> INSERT INTO `test`.`test_cdc` VALUES(1,'zs');

Author: yes, my fault

INSERT INTO `test`.`test_cdc` VALUES(2,'ls');
INSERT INTO `test`.`test_cdc` VALUES(3,'is');
```

Delete from src table, and you will find data also be deleted from target table

```sql
DELETE FROM `test`.`test_cdc` WHERE id = 1 or id = 2;
```

Keypoints
- Delete is only supported in streaming mode, which means delete can only work in MINIBATCH because GLOBAL transaction is for batch mode. If you work in GLOBAL transaction, an exception will be thrown.
- Delete is only supported in upsert mode, If you work in append mode, an exception will be thrown.
- Delete is only supported in tables with at least one pk or valid uk, or an exception will be thrown. valid uk means:
  - The uk's value should not be null.
  - Every column should not be null if uk has multiple-column.

> - Delete compatible with all the codec that is supported by the source, currently including: json craft and canal_json.

Reviewer: ```suggestion
- Delete is compatible with all the codecs supported by the source, currently including: json, craft and canal_json.
```

Author: sorry for the typo

| tidb.source.failover                                     | split         | The TiDB batch stage recovery strategy when it fails, `split` and `offset` are available. `split` recovers from the failed region, guaranteeing at least once semantics, and `offset` recovers from the failed single row data, guaranteeing exactly once semantics. Note that the `offset` behavior may change in higher versions of TiDB (6+), making it impossible to use. |
| tidb.streaming.source                                    | -             | The data source(messaging system) where TiDB's change logs are stored, currently only supports Kafka and Pulsar will be supported later.                                                                                                                                                                                                                                      |

> | tidb.streaming.codec                                     | craft         | TiDB's change log encoding method, currently supports json(called default in the lower version of tidb), craft, canal-json. See [Codec](#7-Codec)                                                                                                                                                                                                                             |

Reviewer: ```suggestion
| tidb.streaming.codec                                     | craft         | TiDB's change log encoding method, currently supports json(called default in the lower version of tidb), craft, canal-json. See [Codec](#8-Codec)                                                                                                                                                                                                                             |
```
